src/invariant/test/ConservationOfLumensTests.cpp

```diff
@@ -301,3 +306,260 @@ TEST_CASE("Inflation changes are consistent",
+                                                   {});    // dead entries
+        ::closeLedger(*app);
+
+        // Create modified version of acc1 with different balance
+        auto acc1Modified = acc1;
+        acc1Modified.data.account().balance = 5000;
+
+        // Add modified acc1 as LIVE entry (shadows the INIT entry)
+        lm.setNextLedgerEntryBatchForBucketTesting(
+            {},             // init entries
+            {acc1Modified}, // live entries (shadows acc1 INIT)
+            {});            // dead entries
+        ::closeLedger(*app);
+
+        // The invariant should only count acc1Modified's balance (5000), not
+        // the original (1000)
+        int64_t expectedTotal =
+            initialTotal + acc1Modified.data.account().balance;
+
+        {
+            LedgerTxn ltx(app->getLedgerTxnRoot());
+            ltx.loadHeader().current().totalCoins = expectedTotal;
+            ltx.commit();
+        }
+
+        ::closeLedger(*app);
+
+        app->getInvariantManager().enableInvariant("ConservationOfLumens");
+
+        auto ledgerState =
+            app->getLedgerManager().getLastClosedLedgerStateForTesting();
+        auto& inMemoryState =
+            app->getLedgerManager().getInMemorySorobanStateForTesting();
+
+        REQUIRE_NOTHROW(app->getInvariantManager().runStateSnapshotInvariant(
+            ledgerState, inMemoryState, []() { return false; }));
+    }
+
+    SECTION("Invariant detects corrupted native balance in hot archive")
+    {
+        // Wrap the existing app in SorobanTest to use contract utilities
+        SorobanTest sorobanTest(app, cfg, true, [](SorobanNetworkConfig&) {});
+
+        auto& root = sorobanTest.getRoot();
+        auto minBalance = app->getLedgerManager().getLastMinBalance(2);
+        auto acc1 = root.create("acc1", minBalance + 10000);
+
+        // Deploy native asset contract and create a balance
+        AssetContractTestClient client(sorobanTest, makeNativeAsset());
+        auto contractAddr = makeContractAddress(sha256("test_contract"));
+
+        // Transfer creates the native balance CONTRACT_DATA entry
+        REQUIRE(client.transfer(acc1, contractAddr, 5000));
+        ::closeLedger(*app);
+
+        // Extract the balance entry from live state
+        LedgerEntry balanceEntry;
+        LedgerKey balanceKey;
+        {
+            LedgerTxn ltx(app->getLedgerTxnRoot());
+            balanceKey = client.makeBalanceKey(contractAddr);
+            auto ltxe = ltx.load(balanceKey);
+            REQUIRE(ltxe);
+            balanceEntry = ltxe.current();
+        }
+
+        // Corrupt: add 123 stroops to the balance value
+        balanceEntry.data.contractData().val.map()->at(0).val.i128().lo += 123;
+
+        // Remove from live buckets and inject corrupted entry into hot archive
+        lm.setNextLedgerEntryBatchForBucketTesting({}, {}, {balanceKey});
+        lm.setNextArchiveBatchForBucketTesting({balanceEntry}, {});
```

Comment:
Nit: for correctness, we should also delete the TTL entry

Reply:
Updated, although let me know if I did that right.